File: backend/core/logging/app_log.py
from __future__ import annotations

import asyncio
import logging
import re
from collections.abc import Mapping
from datetime import UTC, datetime
from typing import Any, Literal
from uuid import uuid4

from pydantic import BaseModel, ConfigDict, Field

AppLogLevel = Literal["debug", "info", "warn", "error", "critical"]
AppLogSource = Literal[
    "backend",
    "frontend",
    "drone",
    "mavlink",
    "telemetry",
    "mission",
    "video",
    "analysis",
    "model",
    "upload",
    "websocket",
    "api",
]

_SENSITIVE_TOKENS = {
    "authorization",
    "cookie",
    "password",
    "passwd",
    "secret",
    "token",
    "access_token",
    "refresh_token",
    "api_key",
    "key",
    "credential",
    "credentials",
}
_MAX_DETAIL_DEPTH = 3
_MAX_STRING_LEN = 500
_MAX_MAPPING_ITEMS = 80
_MAX_SEQUENCE_ITEMS = 50
_BROADCAST_TIMEOUT_S = 1.0

_KEY_SPLIT_RE = re.compile(r"[^a-z0-9]+")
logger = logging.getLogger(__name__)


class AppLogEvent(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str = Field(default_factory=lambda: str(uuid4()))
    timestamp: str = Field(default_factory=lambda: datetime.now(UTC).isoformat())
    level: AppLogLevel
    source: AppLogSource
    message: str = Field(max_length=1000)
    details: dict[str, Any] = Field(default_factory=dict)
    request_id: str | None = None
    mission_id: str | None = None
    flight_id: str | None = None

    def websocket_message(self) -> dict[str, Any]:
        return {"type": "app_log", "data": self.model_dump(mode="json", exclude_none=True)}


def _is_sensitive_key(key: str) -> bool:
    normalized = key.strip().lower()
    if normalized in _SENSITIVE_TOKENS:
        return True
    parts = {part for part in _KEY_SPLIT_RE.split(normalized) if part}
    if parts & _SENSITIVE_TOKENS:
        return True
    return normalized.endswith(("_token", "_secret", "_password", "_api_key", ".token"))


def sanitize_log_details(value: Any, *, _depth: int = 0) -> Any:
    if _depth >= _MAX_DETAIL_DEPTH:
        return str(value)[:_MAX_STRING_LEN]
    if isinstance(value, Mapping):
        sanitized: dict[str, Any] = {}
        for index, (key, raw) in enumerate(value.items()):
            if index >= _MAX_MAPPING_ITEMS:
                sanitized["__truncated__"] = True
                break
            key_s = str(key)
            if _is_sensitive_key(key_s):
                sanitized[key_s] = "[redacted]"
            else:
                sanitized[key_s] = sanitize_log_details(raw, _depth=_depth + 1)
        return sanitized
    if isinstance(value, (list, tuple, set)):
        return [
            sanitize_log_details(item, _depth=_depth + 1)
            for item in list(value)[:_MAX_SEQUENCE_ITEMS]
        ]
    if isinstance(value, bytes):
        return f"[{len(value)} bytes]"
    if isinstance(value, str):
        return value if len(value) <= _MAX_STRING_LEN else f"{value[:_MAX_STRING_LEN]}..."
    if value is None or isinstance(value, (bool, int, float)):
        return value
    return str(value)[:_MAX_STRING_LEN]


async def emit_app_log(
    *,
    level: AppLogLevel,
    source: AppLogSource,
    message: str,
    details: Mapping[str, Any] | None = None,
    request_id: str | None = None,
    mission_id: str | None = None,
    flight_id: str | int | None = None,
) -> AppLogEvent:
    event = AppLogEvent(
        level=level,
        source=source,
        message=str(message)[:1000],
        details=sanitize_log_details(dict(details or {})),
        request_id=request_id,
        mission_id=mission_id,
        flight_id=str(flight_id) if flight_id is not None else None,
    )
    try:
        from backend.infrastructure.messaging.websocket_publisher import telemetry_manager

        await asyncio.wait_for(
            telemetry_manager.broadcast(event.websocket_message()),
            timeout=_BROADCAST_TIMEOUT_S,
        )
    except Exception:
        logger.debug("Failed to broadcast app log event", exc_info=True)
    return event

File: frontend/src/shared/logging/appLog.ts
export type AppLogLevel = "debug" | "info" | "warn" | "error" | "critical";

export type AppLogSource =
  | "backend"
  | "frontend"
  | "drone"
  | "mavlink"
  | "telemetry"
  | "mission"
  | "video"
  | "analysis"
  | "model"
  | "upload"
  | "websocket"
  | "api";

export type AppLogEvent = {
  id: string;
  timestamp: string;
  level: AppLogLevel;
  source: AppLogSource;
  message: string;
  details?: Record<string, unknown>;
  request_id?: string;
  requestId?: string;
  mission_id?: string;
  missionId?: string;
  flight_id?: string;
  flightId?: string;
};

type LogSubscriber = (events: AppLogEvent[]) => void;

const MAX_EVENTS = 150;
const subscribers = new Set<LogSubscriber>();
let events: AppLogEvent[] = [];
const recentKeys = new Map<string, number>();

function nowIso() {
  return new Date().toISOString();
}

function createId() {
  if (typeof crypto !== "undefined" && "randomUUID" in crypto) {
    return crypto.randomUUID();
  }
  return `${Date.now().toString(36)}-${Math.random().toString(36).slice(2)}`;
}

function normalizeLevel(level: AppLogLevel | "warning"): AppLogLevel {
  return level === "warning" ? "warn" : level;
}

function normalizeEvent(input: Omit<AppLogEvent, "id" | "timestamp"> & Partial<Pick<AppLogEvent, "id" | "timestamp">>): AppLogEvent {
  const level = normalizeLevel(input.level as AppLogLevel | "warning");
  return {
    ...input,
    level,
    id: input.id || createId(),
    timestamp: input.timestamp || nowIso(),
    requestId: input.requestId ?? input.request_id,
    missionId: input.missionId ?? input.mission_id,
    flightId: input.flightId ?? input.flight_id,
  };
}

function notify() {
  const snapshot = [...events];
  subscribers.forEach((subscriber) => subscriber(snapshot));
}

function shouldDropDuplicate(event: AppLogEvent) {
  const key = [
    event.level,
    event.source,
    event.message,
    event.requestId ?? event.request_id ?? "",
    event.flightId ?? event.flight_id ?? "",
  ].join("|");
  const now = Date.now();
  for (const [itemKey, seenAt] of recentKeys) {
    if (now - seenAt > 10_000) recentKeys.delete(itemKey);
  }
  const previous = recentKeys.get(key);
  recentKeys.set(key, now);
  return previous !== undefined && now - previous < 10_000;
}

function mirrorToConsole(event: AppLogEvent) {
  const payload = {
    source: event.source,
    requestId: event.requestId ?? event.request_id,
    missionId: event.missionId ?? event.mission_id,
    flightId: event.flightId ?? event.flight_id,
    details: event.details,
  };
  if (event.level === "critical" || event.level === "error") {
    console.error(`[${event.source}] ${event.message}`, payload);
  } else if (event.level === "warn") {
    console.warn(`[${event.source}] ${event.message}`, payload);
  } else if (event.level === "info") {
    console.info(`[${event.source}] ${event.message}`, payload);
  } else {
    console.debug(`[${event.source}] ${event.message}`, payload);
  }
}

export function emitAppLog(
  input: Omit<AppLogEvent, "id" | "timestamp"> & Partial<Pick<AppLogEvent, "id" | "timestamp">>,
  options: { mirrorToConsole?: boolean } = {},
) {
  const event = normalizeEvent(input);
  if (shouldDropDuplicate(event)) return event;
  events = [event, ...events].slice(0, MAX_EVENTS);
  if (options.mirrorToConsole !== false) {
    mirrorToConsole(event);
  }
  notify();
  return event;
}

export function subscribeAppLogs(subscriber: LogSubscriber) {
  subscribers.add(subscriber);
  subscriber([...events]);
  return () => {
    subscribers.delete(subscriber);
  };
}

export function getAppLogs() {
  return [...events];
}

export function clearAppLogsForTests() {
  events = [];
  recentKeys.clear();
  notify();
}

export const frontendLogger = {
  debug: (source: AppLogSource, message: string, details?: Record<string, unknown>) =>
    emitAppLog({ level: "debug", source, message, details }),
  info: (source: AppLogSource, message: string, details?: Record<string, unknown>) =>
    emitAppLog({ level: "info", source, message, details }),
  warn: (source: AppLogSource, message: string, details?: Record<string, unknown>) =>
    emitAppLog({ level: "warn", source, message, details }),
  error: (source: AppLogSource, message: string, details?: Record<string, unknown>) =>
    emitAppLog({ level: "error", source, message, details }),
  critical: (source: AppLogSource, message: string, details?: Record<string, unknown>) =>
    emitAppLog({ level: "critical", source, message, details }),
};
